### DISCUSSION TITLE
Does (or Can) Medusa Support Custom Pricing for Customizable Products?

### DISCUSSION THREAD

**Discussion by burtondev** [medusajs/medusa#8571] [Q&A]
The typical use case scenario is made-to-measure custom blinds, where the customer measures their windows and enters the data, and the price they pay for these customized blinds is then calculated based on the dimensions given. Variant-based fixed pricing is not an option, as this easily adds up to **billions or even trillions of product variants.**

Please see [Case Study: The Bespoke Product Builder That Manages Complex Made-to-Measure Pricing Requirements on Shopify Plus](https://swankyagency.com/bespoke-product-builder-case-study-veneta-blinds/) for a case study of how an agency solved this problem on Shopify. They made an app that generates each custom product on-demand as a completely bespoke product. Otherwise they would have had **744 billion product variants.**

Generating a new product for each order seems like a "hacky" way to do things. Since they are not true "variants" but customizations made by individual customers, these auto-generated products will have to be hidden from product collections, search engines, etc. or deleted.

Does Medusa.js support calculating prices for customizable products on-demand, without using variant-based fixed pricing (which produces billions of variants) and without generating a new product each time (hacky)?

I looked at the [Personalized Products](https://docs.medusajs.com/recipes/personalized-products) "recipe" in the docs and also pricing lists, rules, etc. but I don't think they fit the use case.

I can't find an ecommerce platform that does this out of the box and I don't understand why. With print-on-demand, it's now possible to offer the customer more customizable products, but the ecommerce platforms are holding back. Am I overlooking something? Are there security or legal reasons why this isn't done? I'm at a loss.

Can Medusa.js do this? If so, where do I start?

If it's not supported, are there plans to support it in the near future?

Otherwise, can I either extend Medusa to do this, or to do the "hack" (i.e. set up customizable "products" on the front-end to generate products on the back-end when they are "added to cart"??)

**Comment by ederwii** [comment]
I agree with you in questioning why dynamic pricing for customizable products hasn’t been implemented in most ecommerce platforms yet. However, there are valid reasons for this. Fixed pricing simplifies the system in terms of calculations and reduces the complexity of any formulas that might need to run during the process. Implementing dynamic pricing could introduce complex logic that might not suit every use case, whereas fixed prices work for the vast majority of scenarios, possibly 95% of them.

As a software architect, I can tell you that introducing dynamic pricing comes with potential risks related to data management and performance. The more dynamic the system, the greater the chance for data or performance issues, especially when dealing with a large volume of transactions.

One of the things I appreciate about Medusa is its flexibility. I’m currently working on a system using Medusa as a headless backend, and your discussion brought an interesting point to my attention. Right now, I handle these kinds of “custom” products as manual products by not sending the variant ID during checkout but instead sending the line item data manually. This approach works, but it means the custom item isn’t treated as a true product within the system—just as a custom line item in the order.

I believe Medusa provides enough resources to allow for deep customization. You can create custom pricing logic, custom product configurations, and even custom API endpoints to extend Medusa’s functionality and apply dynamic pricing logic. Of course, it’s worth considering whether this might affect other modules or components, but Medusa’s modular architecture is designed to handle such extensions.

It’s an interesting question, and I’ll definitely revisit this approach when I implement something similar on my end.

**Comment by radoslaw-sz** [comment]
hi @burtondev  - my company handled similar use case commercially. It is possible in Medusa but requires some implementation on the backend and (in some cases) in storefront.

**Comment by docloulou** [comment]
Easiest way would be to create a new product with custom Metadata in it.

**Comment by guyathomas** [comment]
I made a similar post a while ago and received some feedback, but I haven't fully resolved the issue yet. I paused development while waiting for V2 to see what it would bring. [Discussion link](https://github.com/medusajs/medusa/discussions/4679)

@radoslaw-sz, could you share more context on how your company approaches this? I'm considering implementing the following approach:

1. I plan to add middleware for `/store/carts/{id}/line-items` that calls a new dynamic pricing service. This service would create a variant and then include the `variantID` in the request payload. I'll use the metadata key to store attributes for the dynamic pricing service.
2. The dynamic pricing service will assign a PriceSet to a product (rather than a variant, as it's currently implemented), with each price governed by two rules: `min_width` and `min_height`. The service will then select the that matches the constraints of being in the range of the `min_width` and `min_height`.

I'm hoping this will at least get me to the point of adding the line item. I haven’t yet thought through how to handle editing line items, and I’m sure there are other aspects I may have overlooked.

### Pros:
- Preserves the existing constraint that line items require a variant.
- Avoids precomputing billions of variant permutations.
- Should flow through the system correctly.
- Minimal logic overrides are needed, allowing us to keep using the existing APIs.

### Cons:
- The concept of a variant is becoming overloaded. The admin tooling may no longer suit our use case.
- The idea of a product has been changed and we can't simply just query all the variants of it now to render on storefront and need to hide the variants from searches, SSG, SEO, etc...

**Reply by guyathomas** [reply]
Edit: Thinking now, could even make more sense to just have each fabric as a variant and a price_set associated with it. Won't need to create the link then.

**Reply by radoslaw-sz** [reply]
hi @guyathomas  - I cannot give you much details as it was a commercial cooperation, I hope you understand.
From the highest point of view, here are important points which I can mention:
- we created functionality to set your own attributes which influence a price and put them into metadata of variant
- we created editor where you can pick attributes and create how these attributes influence a price. All information were stored in database in separated table and linked to the variant and attributes.
- we exposed new API for adding such "customizable" products (here it was consulted with the customer, that they can handled it separately on the frontend)
- we created a service which takes attributes and based on the define algorithm, calculates dynamically a price

Unfortunately, I cannot give you more details.
